WordPress/playground-tools
# wp-now: execute blueprint steps given a file path

## What?

It adds Blueprint support for wp-now.
Blueprint steps code relies heavily on Web functions and some of them will not work on Javascript for Node.js. We will need to find those steps and convert them into isomorphic javascript.


- Solves https://github.com/WordPress/playground-tools/issues/80
- Related https://github.com/WordPress/playground-tools/issues/9

## Why?



- It will be a great addition to support different configurations such as https://github.com/WordPress/playground-tools/pull/78

## How?



After setting up WordPress and for all the modes except `index` we apply the blueprint steps before any PHP request, so before installing the theme/plugin and before auto-logging.

If the steps `defineWpConfigConsts` or `defineSiteUrl` are passed in the blueprint steps, then we will use its value to change the `absoluteUrl`.

## Testing Instructions

1. Add `127.0.0.1 myurl.wpnow` to your `/etc/hosts`
2. Create a blueprint JSON file in your current directory named `foo.json`.
<details>
<summary>Blueprint example:</summary>

```
{
    "steps": [
        {
            "step": "defineWpConfigConsts",
            "consts": {
                "WP_HOME": "http://myurl.wpnow",
                "WP_SITEURL": "http://myurl.wpnow"
            },
            "virtualize": true
        }
    ]
}
```
Check out more examples: https://wordpress.github.io/wordpress-playground/docs/blueprints-api/examples
</details>

3. Run `nvm use` > `npm install` . **This PR requires Node v20+**
4. Build this branch: `npx nx build wp-now`
5. Run `node dist/packages/wp-now/main.js start --blueprint=foo.json --port=80`
6. Observe `blurpinrt steps: 1` and `Blueprint step completed: defineWpConfigConsts` appear in the console output
7. Observe the URL `http://myurl.wpnow` opens in your browser and WordPress still working as expected.


Ngrok case

1. Run `ngrok http 8881`.
2. Modify the `foo.json` file with the ngrok temporal URL.
3. Run `node dist/packages/wp-now/main.js start --blueprint=foo.json --port=80`
4. Observe the ngrok URL opens and WordPress works as expected.
